addons/hr_holidays/models/hr_leave.py

```diff
@@ -353,15 +353,25 @@ def _compute_from_holiday_status_id(self):
         for leave in self:
             if leave.holiday_status_id.requires_allocation == 'yes' and leave.date_from and leave.date_to:
                 found_allocation = False
+                number_of_days = []
                 date_to = leave.date_to.replace(tzinfo=UTC).astimezone(timezone(leave.tz)).date()
                 date_from = leave.date_from.replace(tzinfo=UTC).astimezone(timezone(leave.tz)).date()
                 for allocation in allocations_dict[(leave.holiday_status_id.id, leave.employee_id.id)]:
                     date_to_check = allocation['date_to'] >= date_to if allocation['date_to'] else True
                     date_from_check = allocation['date_from'] <= date_from
                     if (date_to_check and date_from_check):
-                        found_allocation = allocation['id']
-                        break
-                leave.holiday_allocation_id = self.env['hr.leave.allocation'].browse(found_allocation) if found_allocation else False
+                        if allocation['number_of_days'] >= self.number_of_days:
+                            found_allocation = allocation['id']
+                            break
+                        number_of_days.append(allocation['number_of_days'])
+                if found_allocation:
+                    leave.holiday_allocation_id = self.env['hr.leave.allocation'].browse(found_allocation)
+                else:
+                    if sum(number_of_days) >= self.number_of_days:
+                        raise UserError(("You have %g allocations for those type and period. They have respectively %s days. "
+                        + "Please split your request to fit in their number of days.") % (len(number_of_days), number_of_days))
```

Comment:
The string needs to be translated and added to the .pot file 
https://www.odoo.com/documentation/15.0/developer/misc/i18n/translations.html
```suggestion
                        raise UserError(_("You have %(count)s allocations for those type and period. They have respectively %(count)s days. " +\
                        "Please split your request to fit in their number of days.", count=number_of_days)
```